src/bin/async_pipe.rs

```diff
@@ -0,0 +1,65 @@
+use async_trait::async_trait;
+use raw_ipa::build_async_pipeline;
+use raw_ipa::error::{Error, Res};
+use raw_ipa::pipeline::async_pipe::{APipeline, AStep};
+use std::time::Duration;
+
+/// unchanged from regular pipeline
+struct Start {
+    x: i32,
+    y: i32,
+}
+#[async_trait]
+impl AStep for Start {
+    type Input = ();
+    type Output = (i32, i32);
+
+    async fn compute(&self, _: Self::Input) -> Res<Self::Output> {
+        Ok((self.x, self.y))
+    }
+}
+
+/// unchanged from regular pipeline
+struct Add {}
+#[async_trait]
+impl AStep for Add {
+    type Input = (i32, i32);
+    type Output = i32;
+
+    async fn compute(&self, inp: Self::Input) -> Res<Self::Output> {
+        Ok(inp.0 + inp.1)
+    }
+}
+
+/// arbitrary async work done (literally a `time::sleep`) to prove that it can occur
+struct PairWith3 {}
+#[async_trait]
+impl AStep for PairWith3 {
+    type Input = i32;
+    type Output = (i32, i32);
+
+    async fn compute(&self, inp: Self::Input) -> Res<Self::Output> {
+        let (send, recv) = tokio::sync::oneshot::channel::<i32>();
+        tokio::spawn(async move {
```

Comment:
I believe you can just await task handle her

```
       async fn compute(&self, inp: Self::Input) -> Res<Self::Output> {
        let res = tokio::spawn(async move {
            tokio::time::sleep(Duration::from_millis(500)).await;
            3
        }).await.map_err(...)?;
        
        Ok((inp, res))
    }

```

Reply:
yeah. same effect, but no channels. switched to that